feat(appeng): integrate StackApi for full fluid and essentia support.

hinyb opened the pull request:
This PR refactors the old logic via new StackAPI, providing full support for fluid and essentia, matching the esisting item support.

Breaking Changes
To maintain consistency across the API, update the parameter order of `setPatternSlot` to align with `storeInterfacePattern` and `clearInterfacePattern`

- Old: `function(slot:number, database:address, entry:number, size:number, index:number)`
- New: `function(slot:number, index:number[, database:address, entry:number, size:number]):boolean`

Changes:
- Full fluid and essentia Support via the new StackAPI.
- Add network_changed event subscription.
- Fix thread-safety issues in allItems.
- Add integration for ae2fc.
- Add integration for Storage Bus and OreFilter.

fix https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/19874
fix https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/19567
fix https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/20839
fix https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/23349
Fix https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/23718
fix https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/23948

hinyb committed:
feat(appeng): add allFluids and network_changed event subscription.

hinyb committed:
fix: resolve thread-safety issues in allItems and allFluids.

hinyb commented:
fix https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/19874 https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/19567 https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/20839

hinyb committed:
feat(appeng): add fluid support.

hinyb committed:
Merge branch 'master' into appeng

hinyb committed:
fix: fluid stacks and NetworkContents failing to persist.

konradmoesch commented:
I have tested the build artifact and my crashes have gone away, no new issues.
Whats still needed to be done? Can I help w/ anything? Looks good to me

hinyb committed:
feat: full fluid and essentia support via new Stack Type API.
BREAKING CHANGE: update the parameter order of `setPatternSlot` to align with `storeInterfacePattern` and `clearInterfacePattern`

hinyb committed:
refactor: merge setPartConfig and Extended into single setPartConfig for API consistency.

hinyb committed:
build: remove unnecessary runtimeOnly dependency

hinyb committed:
fix: add missing wildcard case to prevent MatchError.

hinyb committed:
refactor: clean up AE2 traits and fix type matching.

hinyb committed:
feat: add support for AE2FC blocks.

hinyb committed:
feat: add support for ore filter and storage bus.

hinyb committed:
fix: subscription persistence and remove thread-unsafe allFluids and allEssentia.
- perf: use view for AE2 storage to reduce heap allocation.

hinyb committed:
refactor: improve subscription logic using EventHandler

hinyb committed:
fix: prevent oc from voiding requested items

hinyb committed:
feat: add id support for getItemInNetwork

hinyb committed:
refactor: update activeItems, pendingItems and storedItems

Dream-Master committed:
Merge branch 'master' into appeng

hinyb committed:
fix: support tags in filter

Dream-Master committed:
Merge branch 'master' into appeng

Kogepan229 reviewed (commented):
@hinyb 
Thank you for fixing the compatibility!
I'm not familiar with Scala or Overclocking, but I found a few things that caught my attention.

hinyb committed:
fix: correct type in setExportConfiguration

Co-authored-by: Kogepan229 <73163994+Kogepan229@users.noreply.github.com>

hinyb committed:
fix: correct logic for saving links

hinyb committed:
fix: add missing imports

Kogepan229 reviewed (approved):
Thank you very much!
I can't say it's perfectly reviewed, but since there doesn't seem to be anyone else who can review, let's merge and test.

unilock commented:
It appears that all versions of OpenComputers containing this PR crash when AE2 is not installed. v1.12.22 is the last (stable) version that does not include this PR, and thus, does not crash.

<details>

<summary>ClassNotFoundException as of v1.12.38</summary>

```
[10:11:58] [Client thread/WARN] [OpenComputers]: Something went wrong!
java.lang.NoClassDefFoundError: appeng/api/storage/data/IAEStackType
	at Launch//li.cil.oc.integration.Mods$.<init>(Mods.scala:95) ~[Mods$.class:?]
	at Launch//li.cil.oc.integration.Mods$.<clinit>(Mods.scala) ~[Mods$.class:?]
	at Launch//li.cil.oc.integration.Mods.All(Mods.scala) ~[Mods.class:?]
	at Launch//li.cil.oc.common.asm.TransformerInjectInterfaces.injectInterface(TransformerInjectInterfaces.java:72) ~[OpenComputers-1.12.38-GTNH.jar:?]
	at Launch//li.cil.oc.common.asm.TransformerInjectInterfaces.transform(TransformerInjectInterfaces.java:35) ~[OpenComputers-1.12.38-GTNH.jar:?]
	at Launch//li.cil.oc.common.asm.ClassTransformer.transform(ClassTransformer.java:61) [OpenComputers-1.12.38-GTNH.jar:?]
	at System//net.minecraft.launchwrapper.LaunchClassLoader.runTransformers(LaunchClassLoader.java:409) [LaunchClassLoader.class:?]
	at System//net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:293) [LaunchClassLoader.class:?]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:557) [?:?]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:490) [?:?]
	at java.base/java.lang.ClassLoader.defineClass1(Native Method) [?:?]
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:962) [?:?]
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:144) [?:?]
	at System//net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:334) [LaunchClassLoader.class:?]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:557) [?:?]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:490) [?:?]
	at java.base/java.lang.ClassLoader.defineClass1(Native Method) [?:?]
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:962) [?:?]
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:144) [?:?]
	at System//net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:334) [LaunchClassLoader.class:?]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:557) [?:?]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:490) [?:?]
	at java.base/java.lang.ClassLoader.defineClass1(Native Method) [?:?]
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:962) [?:?]
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:144) [?:?]
	at System//net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:334) [LaunchClassLoader.class:?]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:557) [?:?]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:490) [?:?]
	at Launch//com.mitchej123.hodgepodge.core.rfb.hooks.ForgeEventSubscriptionHook.shouldTransformClass(ForgeEventSubscriptionHook.java:14) [hodgepodge-2.7.134.jar:?]
	at Launch//cpw.mods.fml.common.asm.transformers.EventSubscriptionTransformer.transform(EventSubscriptionTransformer.java:61) [forge-1.7.10-10.13.4.1614-1.7.10-universal.jar:?]
	at System//net.minecraft.launchwrapper.LaunchClassLoader.runTransformers(LaunchClassLoader.java:409) [LaunchClassLoader.class:?]
	at System//net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:293) [LaunchClassLoader.class:?]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:557) [?:?]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:490) [?:?]
	at Launch//li.cil.oc.common.init.Blocks$.init(Blocks.scala:13) [Blocks$.class:?]
	at Launch//li.cil.oc.common.Proxy.preInit(Proxy.scala:36) [Proxy.class:?]
	at Launch//li.cil.oc.client.Proxy.preInit(Proxy.scala:37) [Proxy.class:?]
	at Launch//li.cil.oc.OpenComputers$.preInit(OpenComputers.scala:41) [OpenComputers$.class:?]
	at Launch//li.cil.oc.OpenComputers.preInit(OpenComputers.scala) [OpenComputers.class:?]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104) ~[?:?]
	at java.base/java.lang.reflect.Method.invoke(Method.java:565) ~[?:?]
	at Launch//cpw.mods.fml.common.FMLModContainer.handleModStateEvent(FMLModContainer.java:532) [FMLModContainer.class:?]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104) ~[?:?]
	at java.base/java.lang.reflect.Method.invoke(Method.java:565) ~[?:?]
	at System//com.google.common.eventbus.EventSubscriber.handleEvent(EventSubscriber.java:74) [guava-17.0.jar:?]
	at System//com.google.common.eventbus.SynchronizedEventSubscriber.handleEvent(SynchronizedEventSubscriber.java:47) [guava-17.0.jar:?]
	at System//com.google.common.eventbus.EventBus.dispatch(EventBus.java:322) [guava-17.0.jar:?]
	at System//com.google.common.eventbus.EventBus.dispatchQueuedEvents(EventBus.java:304) [guava-17.0.jar:?]
	at System//com.google.common.eventbus.EventBus.post(EventBus.java:275) [guava-17.0.jar:?]
	at Launch//cpw.mods.fml.common.LoadController.sendEventToModContainer(LoadController.java:212) [LoadController.class:?]
	at Launch//cpw.mods.fml.common.LoadController.propogateStateMessage(LoadController.java:190) [LoadController.class:?]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104) ~[?:?]
	at java.base/java.lang.reflect.Method.invoke(Method.java:565) ~[?:?]
	at System//com.google.common.eventbus.EventSubscriber.handleEvent(EventSubscriber.java:74) [guava-17.0.jar:?]
	at System//com.google.common.eventbus.SynchronizedEventSubscriber.handleEvent(SynchronizedEventSubscriber.java:47) [guava-17.0.jar:?]
	at System//com.google.common.eventbus.EventBus.dispatch(EventBus.java:322) [guava-17.0.jar:?]
	at System//com.google.common.eventbus.EventBus.dispatchQueuedEvents(EventBus.java:304) [guava-17.0.jar:?]
	at System//com.google.common.eventbus.EventBus.post(EventBus.java:275) [guava-17.0.jar:?]
	at Launch//cpw.mods.fml.common.LoadController.distributeStateMessage(LoadController.java:119) [LoadController.class:?]
	at Launch//cpw.mods.fml.common.Loader.preinitializeMods(Loader.java:556) [Loader.class:?]
	at Launch//cpw.mods.fml.client.FMLClientHandler.beginMinecraftLoading(FMLClientHandler.java:243) [FMLClientHandler.class:?]
	at Launch//net.minecraft.client.Minecraft.func_71384_a(Minecraft.java:480) [bao.class:?]
	at Launch//net.minecraft.client.Minecraft.func_99999_d(Minecraft.java:9599) [bao.class:?]
	at Launch//net.minecraft.client.main.Main.main(SourceFile:148) [Main.class:?]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104) ~[?:?]
	at java.base/java.lang.reflect.Method.invoke(Method.java:565) ~[?:?]
	at System//net.minecraft.launchwrapper.Launch.rfb$realLaunch(Launch.java:250) [Launch.class:?]
	at System//net.minecraft.launchwrapper.Launch.launch(Launch.java:35) [Launch.class:?]
	at System//net.minecraft.launchwrapper.Launch.main(Launch.java:60) [Launch.class:?]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104) ~[?:?]
	at java.base/java.lang.reflect.Method.invoke(Method.java:565) ~[?:?]
	at com.gtnewhorizons.retrofuturabootstrap.Main.main(Main.java:207) [lwjgl3ify-3.0.16-forgePatches.jar:3.0.16]
	at com.gtnewhorizons.retrofuturabootstrap.MainStartOnFirstThread.lambda$main$1(MainStartOnFirstThread.java:47) [lwjgl3ify-3.0.16-forgePatches.jar:3.0.16]
	at java.base/java.lang.Thread.run(Thread.java:1474) [?:?]
Caused by: java.lang.ClassNotFoundException: Class bytes are null for appeng.api.storage.data.IAEStackType (appeng.api.storage.data.IAEStackType, appeng.api.storage.data.IAEStackType)
	at System//net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:326) ~[LaunchClassLoader.class:?]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:557) ~[?:?]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:490) ~[?:?]
	... 74 more
```

</details>
